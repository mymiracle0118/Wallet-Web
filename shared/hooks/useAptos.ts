import create from 'zustand'
import { devtools, persist } from 'zustand/middleware'
import { Ed25519Keypair } from '@mysten/sui.js/keypairs/ed25519'
import { chromeSyncStorage } from '@portal/shared/utils/chromeSyncStorage'
import { getAptosWalletUsingSeed, getBalance } from '@portal/shared/services/aptosService'
import { ethers } from 'ethers'
import { toUTF8Array } from '../utils/utf8'

type APTOSWalletState = {
  /** APTOS network - create wallet */
  createAptosWallet: (mnemonics?: string) => Promise<any>
  /** APTOS network - Update balance */
  refreshBalance: (aptosType?: string) => Promise<void>
  _keypair?: Ed25519Keypair | null
  /** APTOS network - Wallet Address */
  address?: string
  /** APTOS network - privateKey Address */
  privateKey?: string
  /** APTOS network - Total Balance of wallet */
  balance: number
  /** APTOS network - All objects owned by wallet */
  objectsOwned?: Array<any>

  wallet?: ethers.Wallet
  encryptedPrivateKey: (string | number)[]
  encryptedWallet?: string
}

const initialState = {
  address: undefined,
  balance: 0,
  _keypair: undefined,
  privateKey: undefined,
  encryptedPrivateKey: [],
}

export const useAptos = create<APTOSWalletState>()(
  devtools(
    persist(
      (set, get) => ({
        ...initialState,

        createAptosWallet: async (mnemonics?: string) => {
          const wallet = await getAptosWalletUsingSeed(mnemonics)
          if (wallet) {
            const encryptedPrivateKey = toUTF8Array(wallet?.privateKey)
            set({ address: wallet?.address, encryptedPrivateKey })
          }

          const { refreshBalance } = get()
          await refreshBalance()
          return wallet
        },
        refreshBalance: async (aptosType?: string) => {
          const { address } = get()
          if (!address) return
          const balance = await getBalance(address, aptosType)
          set({ balance: balance })
        },
      }),
      {
        name: 'APTOS-storage',
        getStorage: () => chromeSyncStorage,
        partialize: (state) => Object.fromEntries(Object.entries(state).filter(([key]) => !['wallet'].includes(key))), // Prevents decrypted wallet from being persisted in storage
      }
    ),
    { name: 'APTOS - Wallet' }
  )
)
