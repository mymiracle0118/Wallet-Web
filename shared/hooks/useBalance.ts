import { useEffect, useMemo } from 'react'
import { useQuery } from '@tanstack/react-query'
import { BigNumber } from 'ethers'

import { NetworkFactory } from '@portal/shared/factory/network.factory'
import { NetworkToken } from '../utils/types'
import { useStore } from './useStore'

type BalanceResponse = {
  balance: BigNumber | number
  balanceFormatted: number
  isLoading: boolean
}
const { updateTokenBalance } = useStore.getState()
export const useBalance = (token: NetworkToken): BalanceResponse => {
  let balance, balanceFormatted

  const networkFactory = NetworkFactory.selectByNetworkId(token.shortName)
  if (!networkFactory) {
    return {
      balance: 0,
      balanceFormatted: 0,
      isLoading: false,
    }
  }

  const walletAddress = token.address
  const { data, isLoading } = useQuery(
    ['asset-balance', token.networkName, walletAddress, token.shortName],
    () => networkFactory.getTokenBalance(walletAddress as string, token),
    {
      refetchInterval: 60000,
      staleTime: 30000,
    }
  )

  balance = useMemo(() => {
    return data || BigNumber.from(0)
  }, [data])
  balanceFormatted = useMemo(() => networkFactory.getBalanceFormattedValue(data), [data]) as number

  useEffect(() => {
    // Update asset balance in wallet to keep running total
    if (balance && balanceFormatted) {
      updateTokenBalance(token.shortName, balance, balanceFormatted)
    }
  }, [balance, balanceFormatted])

  return {
    balance,
    balanceFormatted,
    isLoading,
  }
}
