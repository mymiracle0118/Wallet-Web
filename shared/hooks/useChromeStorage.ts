import { useCallback, useEffect, useState } from 'react'

export const storage = {
  get: (key: string, defaultValue: any, storageArea: Storage) => {
    const keyObj = defaultValue === undefined ? key : { [key]: defaultValue }
    return new Promise((resolve, reject) => {
      chrome.storage[storageArea].get(keyObj, (items: any) => {
        const error = chrome.runtime.lastError
        if (error) return reject(error)
        resolve(items[key])
      })
    })
  },
  set: (key: string, value: any, storageArea: Storage) => {
    return new Promise((resolve, reject) => {
      chrome.storage[storageArea].set({ [key]: value }, () => {
        const error = chrome.runtime.lastError
        error ? reject(error) : resolve()
      })
    })
  },
}

/**
 * Basic hook for storage
 * @param {string} key
 * @param {*} initialValue
 * @param {'local'|'sync'} storageArea
 * @returns {[*, function(*= any): void, boolean, string]}
 */

export default function useChromeStorage(key, initialValue, storageArea) {
  const [INITIAL_VALUE] = useState(() => {
    return typeof initialValue === 'function' ? initialValue() : initialValue
  })
  const [STORAGE_AREA] = useState(storageArea)
  const [state, setState] = useState(INITIAL_VALUE)
  const [isPersistent, setIsPersistent] = useState(true)
  const [error, setError] = useState('')

  useEffect(() => {
    storage
      .get(key, INITIAL_VALUE, STORAGE_AREA)
      .then((res) => {
        setState(res)
        setIsPersistent(true)
        setError('')
      })
      .catch((error) => {
        setIsPersistent(false)
        setError(error)
      })
  }, [key, INITIAL_VALUE, STORAGE_AREA])

  const updateValue = useCallback(
    (newValue) => {
      const toStore = typeof newValue === 'function' ? newValue(state) : newValue
      storage
        .set(key, toStore, STORAGE_AREA)
        .then(() => {
          setIsPersistent(true)
          setError('')
        })
        .catch((error) => {
          // set newValue to local state because chrome.storage.onChanged won't be fired in error case
          setState(toStore)
          setIsPersistent(false)
          setError(error)
        })
    },
    [STORAGE_AREA, key, state]
  )

  useEffect(() => {
    const onChange = (changes, areaName) => {
      if (areaName === STORAGE_AREA && key in changes) {
        setState(changes[key].newValue)
        setIsPersistent(true)
        setError('')
      }
    }
    chrome.storage.onChanged.addListener(onChange)
    return () => {
      chrome.storage.onChanged.removeListener(onChange)
    }
  }, [key, STORAGE_AREA])

  return [state, updateValue, isPersistent, error]
}
