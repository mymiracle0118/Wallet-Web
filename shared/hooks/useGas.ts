import React, { useEffect, useState } from 'react'
import { useQuery } from '@tanstack/react-query'
import { ethers } from 'ethers'
import { getContractInterface } from '@portal/shared/utils/getContractInterface'
import { NetworkFactory } from '@portal/shared/factory/network.factory'
import { Asset, GasOption, NetworkToken } from '@portal/shared/utils/types'

const getERC20GasLimit = async (asset: Asset, provider: any): Promise<number> => {
  if (asset.type === 'erc721') {
    return 100000
  }
  const contract = new ethers.Contract(asset.contractAddress!, getContractInterface(asset.type), provider)
  const deploymentData = contract.interface.encodeDeploy([])
  const estimatedGasLimit = await provider.estimateGas({ data: deploymentData })
  return Number(ethers.utils.formatUnits(estimatedGasLimit, 'wei'))
}

export const useGas = (asset: NetworkToken, option?: GasOption) => {
  const [gasLimit, setGasLimit] = useState(21000)
  const [gasOption, setGasOption] = useState(option || GasOption.Aggressive)
  const getNetwork = NetworkFactory.selectByNetworkId(asset.networkName)
  const provider = getNetwork.provider
  const { data: feeData } = useQuery(
    ['gas-fee-data', asset],
    () => {
      return getNetwork.getFeeData()
    },
    {
      staleTime: 15000,
      refetchInterval: 15000,
    }
  )

  const { data: gasLimitData } = useQuery(
    ['gas-limit-data', asset, provider],
    () => getERC20GasLimit(asset, provider),
    {
      enabled: asset.tokenContractAddress !== undefined,
      staleTime: 60 * 60000,
    }
  )

  useEffect(() => {
    if (gasLimitData) {
      setGasLimit(gasLimitData)
    }
  }, [gasLimitData])

  const maxFeePerGas = React.useMemo(() => {
    if (!feeData || feeData.gasPrice === null) {
      return ethers.utils.parseUnits('0', 'gwei')
    }

    let baseGas = parseFloat(ethers.utils.formatUnits(feeData.gasPrice || 0, 'gwei'))
    const baseGasPlusMultiplier = (multiplier: number) => {
      return baseGas * multiplier
    }

    switch (gasOption) {
      case GasOption.Low:
        baseGas = baseGasPlusMultiplier(1.0)
        break
      case GasOption.Market:
        baseGas = baseGasPlusMultiplier(1.5)
        break
      case GasOption.Aggressive:
        baseGas = baseGasPlusMultiplier(2.0)
        break
    }

    return ethers.utils.parseUnits(baseGas.toFixed(8), 'gwei')
  }, [gasOption, feeData])

  const maxPriorityFeePerGas = React.useMemo(() => {
    return getNetwork.getMaxPriorityFeePerGas({ maxFeePerGas, gasOption })
  }, [gasOption, maxFeePerGas])

  const marketGas = React.useMemo(async () => {
    return await getNetwork.marketGas(maxFeePerGas)
  }, [gasOption, maxFeePerGas])

  const estimatedTransactionCost = React.useMemo(async () => {
    return maxFeePerGas && (await getNetwork.getEstimatedTransactionCost(maxFeePerGas, maxPriorityFeePerGas, gasLimit))
  }, [gasLimit, gasOption, maxFeePerGas])

  const estimatedTransactionTime = React.useMemo(() => {
    switch (gasOption) {
      case GasOption.Low:
        return '30 seconds'
      case GasOption.Market:
        return '30 seconds'
      case GasOption.Aggressive:
        return '15 seconds'
    }
  }, [gasOption])

  return {
    gasLimit,
    gasOption,
    setGasOption,
    marketGas,
    maxFeePerGas,
    maxPriorityFeePerGas,
    estimatedTransactionCost,
    estimatedTransactionTime,
  }
}
