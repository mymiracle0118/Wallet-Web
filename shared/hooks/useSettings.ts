import create from 'zustand'
import { devtools, persist } from 'zustand/middleware'
import produce from 'immer'

import type { Asset } from './useWallet'
import { AccountWallet, useWallet } from './useWallet'
import { SAFETY_MEASURE } from '@portal/portal-extension/src/utils/constants'
import { chromeSyncStorage } from '@portal/shared/utils/chromeSyncStorage'
import { setLockUp } from './useLockUp'
import { EnvironmentType, NetworkToken } from '@portal/shared/utils/types'
import { useStore } from './useStore'

export type AddressBookUser = {
  avatar?: string
  username: string
  address: string
  safety?: SAFETY_MEASURE
  network: string
}
export type AccountTokenType = {
  isFavorite: boolean
}
export type TAccount = {
  assets: Array<Asset>
  address: string
  username: string
  encryptedPrivateKey: string
  encryptedWallet: string
  avatar: string
  balance: number
  isAccountImported: boolean
  networkName: string
  tokens: { [env: string]: { [key: string]: AccountTokenType } }
  customTokens: { [env: string]: { [key: string]: NetworkToken } }
  isPrimary: boolean
}

type TAddressBook = {
  username: string
  address: string
}

export type SettingsState = {
  enablePasswordProtection: boolean
  accounts: { [address: string]: TAccount }
  currentAccount: TAccount | null
  darkMode: boolean
  addressBook: AddressBookUser[]
  saveAccount: (
    username: string,
    address: string,
    networkName: string,
    isPrimary?: boolean,
    isAccountImported?: boolean
  ) => void
  clearAccounts: () => void
  clearAddressbook: () => void
  setDarkMode: (mode: boolean) => void
  setEnablePasswordProtection: (enable: boolean) => void
  hideAccount: (address: string) => void
  addToAddressBook: (user: AddressBookUser) => void
  removeToAddressBook: (address: string) => void
  updateUsername: (username: string, address: string) => void
  networkEnvironment: EnvironmentType
  setNetworkEnvironment: (envType: EnvironmentType) => void
  setCurrentAccount: (account: TAccount) => void
  addToken: (accountKey: string, tokenKey: string) => void
  updateTokens: (accountKey: string, tokenKeys: string[]) => void
  addRemoveFavoriteToken: (accountKey: string, tokenKey: string, isFavoriteValue: boolean) => void
  addCustomToken: (token: NetworkToken) => void
}

export const useSettings = create<SettingsState>()(
  devtools(
    persist(
      (set, get) => ({
        accounts: {},
        currentAccount: null,
        addressBook: [],
        enablePasswordProtection: false,
        darkMode: true,
        networkEnvironment: 'testNet',
        setDarkMode: (mode) => {
          set({ darkMode: mode })
        },

        updateUsername: async (newUsername: string, address: string) => {
          set(
            produce((state: SettingsState) => {
              state.accounts[address].username = newUsername
            })
          )
        },

        addToAddressBook: (user: AddressBookUser) => {
          set(
            produce((state: SettingsState) => ({
              addressBook: [...state.addressBook, user],
            }))
          )
        },

        removeToAddressBook: (address: string) => {
          set(
            produce((state: SettingsState) => {
              state.addressBook = state.addressBook.filter((v) => v.address !== address)
            })
          )
        },

        saveAccount: async (username, address, networkName, isPrimary, isAccountImported) => {
          const { avatar, encryptedWallet } = useWallet.getState()
          const { addTokenToList } = useStore.getState()
          if (username && address) {
            const account = {
              address,
              username,
              encryptedWallet,
              avatar,
              networkName,
              isAccountImported: isAccountImported ? isAccountImported : false,
              tokens: { testNet: {}, mainNet: {} },
              customTokens: { testNet: {}, mainNet: {} },
              isPrimary: isPrimary || false,
            }
            set(
              produce((state: SettingsState) => {
                // @ts-ignore
                state.accounts[address] = account
              })
            )
            const { accounts, currentAccount } = get()
            if (!currentAccount || (currentAccount && !currentAccount.address)) {
              set({ currentAccount: accounts[address] })
            }
            addTokenToList(address, networkName)
          }
        },

        clearAccounts: () => {
          set({ accounts: {}, currentAccount: null })
        },

        clearAddressbook: () => {
          set({ addressBook: [] })
        },

        setEnablePasswordProtection: (enable: boolean) => {
          setLockUp(enable)
          set({
            enablePasswordProtection: enable,
          })
        },

        hideAccount: (addressHide: string) => {
          const accountLists = { ...get().accounts }
          if (accountLists[`${addressHide}`]) {
            delete accountLists[`${addressHide}`]
          }

          set(
            produce((state: SettingsState) => {
              state.accounts = accountLists
            })
          )

          const { walletsList } = useStore.getState()
          if (walletsList[`${addressHide}`]) {
            delete walletsList[`${addressHide}`]
            useStore.getState().removeWalletFromWalletList(walletsList)
          }

          const { address } = useWallet.getState()
          if (address) {
            const account = useSettings.getState().accounts[address]
            useSettings.getState().setCurrentAccount(account)
          }
        },
        setNetworkEnvironment: (envType: EnvironmentType) => {
          set(
            produce((state: SettingsState) => {
              state.networkEnvironment = envType
            })
          )
        },
        setCurrentAccount: (account: TAccount) => {
          const { updateCurrentAccountTokensList } = useStore.getState()
          set(
            produce((state: SettingsState) => {
              state.currentAccount = account
            })
          )
          updateCurrentAccountTokensList(account.address)
        },
        addToken: (accountKey, tokenKey: string) => {
          const { networkEnvironment } = get()
          set(
            produce((state: SettingsState) => {
              const existingToken = { ...state.accounts[accountKey].tokens[networkEnvironment] }
              const newToken: AccountTokenType = { isFavorite: false }
              existingToken[tokenKey] = newToken
              state.accounts[accountKey].tokens[networkEnvironment] = existingToken
            })
          )
        },
        addRemoveFavoriteToken: (accountKey: string, tokenKey: string, isFavoriteValue: boolean) => {
          const { networkEnvironment } = get()
          set(
            produce((state: SettingsState) => {
              const existingToken = { ...state.accounts[accountKey].tokens[networkEnvironment] }
              const newToken: AccountTokenType = { isFavorite: false }
              existingToken[tokenKey] = { ...existingToken[tokenKey], isFavorite: isFavoriteValue }
              state.accounts[accountKey].tokens[networkEnvironment] = existingToken
            })
          )
        },
        updateTokens: (accountKey: string, tokenKeys: string[]) => {
          const { networkEnvironment } = get()
          set(
            produce((state: SettingsState) => {
              const existingTokens = { ...state.accounts[accountKey].tokens[networkEnvironment] }
              for (const key in existingTokens) {
                if (!tokenKeys.includes(key)) {
                  delete existingTokens[key]
                }
              }
              state.accounts[accountKey].tokens[networkEnvironment] = existingTokens
            })
          )
        },
        addCustomToken: async (token: NetworkToken) => {
          const { networkEnvironment, currentAccount } = get()
          if (currentAccount) {
            const accountKey = currentAccount.address
            set(
              produce((state: SettingsState) => {
                if (!state.accounts[accountKey].customTokens[networkEnvironment]) {
                  state.accounts[accountKey].customTokens[networkEnvironment] = {}
                }
                const existingToken = { ...state.accounts[accountKey].customTokens[networkEnvironment] }
                existingToken[token.shortName] = token
                state.accounts[accountKey].customTokens[networkEnvironment] = existingToken
              })
            )

            const { addTokenToList, addWalletToList, walletsList } = useStore.getState()

            const newWalletObj: AccountWallet = { ...walletsList[accountKey]['ETH'] }
            await addWalletToList(accountKey, token.networkName, newWalletObj)
            addTokenToList(accountKey, token.shortName)
          }
        },
      }),
      {
        name: 'account-storage',
        getStorage: () => chromeSyncStorage,
      }
    ),
    { name: 'Shuttle - Accounts' }
  )
)
