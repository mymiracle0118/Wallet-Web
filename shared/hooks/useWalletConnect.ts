import create from 'zustand'
import { devtools, persist } from 'zustand/middleware'
import type { IClientMeta, IWalletConnectSession } from '@walletconnect/legacy-types'
import WalletConnect from '@walletconnect/legacy-client'
import { ethers } from 'ethers'

import { chromeSyncStorage } from '@portal/shared/utils/chromeSyncStorage'
import { getNetworkProviderByChainId } from '@portal/shared/utils/getNetworkProvider'
import { useSettings } from './useSettings'

export type SessionRequest = {
  id: number
  method: string
  params: Array<{ peerId: string; peerMeta: IClientMeta; chainId: number }>
}

export type CallRequest = {
  id: number
  method: string
  params: Array<unknown>
}

export type WalletConnectState = {
  connector?: WalletConnect
  sessionRequest?: SessionRequest
  callRequest?: CallRequest
  session?: IWalletConnectSession
  selectedChainId: number
  createSession: (uri?: string) => void
  approveSession: (accounts: Array<string>) => void
  rejectSession: () => void
  approveCallRequest: (walletSigner: ethers.Wallet) => Promise<void>
  rejectCallRequest: () => void
  setSelectedChainId: (chainId: number) => void
  disconnect: () => void
  clear: () => void
}

const initialState = {
  selectedChainId: 1,
  connector: undefined,
  sessionRequest: undefined,
  callRequest: undefined,
  session: undefined,
}

const clientMeta = {
  description: 'Shuttle Wallet Developer App',
  url: 'https://walletconnect.org',
  icons: ['https://walletconnect.org/walletconnect-logo.png'],
  name: 'Shuttle Wallet',
}

export const useWalletConnect = create<WalletConnectState>()(
  devtools(
    persist(
      (set, get) => ({
        ...initialState,
        createSession: (uri?: string) => {
          let connector: WalletConnect

          const { session, selectedChainId } = get()
          if (session && session.connected) {
            connector = new WalletConnect({ session })
            if (selectedChainId !== 1) {
              connector.updateSession({ accounts: session.accounts, chainId: selectedChainId })
            }
          } else {
            connector = new WalletConnect({ uri, clientMeta })
          }

          set({ connector, session: connector.session })

          connector.on('session_request', (error, payload) => {
            set({ sessionRequest: payload })
          })

          connector.on('call_request', (error, payload) => {
            set({ callRequest: payload })
          })

          connector.on('disconnect', (error, payload) => {
            set(initialState)
          })
        },

        approveSession: (accounts: Array<string>) => {
          const { sessionRequest, connector, selectedChainId } = get()
          if (!sessionRequest) {
            throw new Error('No session request')
          }

          if (!connector) {
            throw new Error('No connector')
          }

          connector.approveSession({ accounts, chainId: selectedChainId })
          set({ sessionRequest: undefined, session: connector.session })
        },

        rejectSession: () => {
          const { connector, sessionRequest } = get()
          if (!sessionRequest) {
            throw new Error('No session request')
          }

          if (!connector) {
            throw new Error('No connector')
          }

          connector.rejectSession()
          set(initialState)
        },

        approveCallRequest: async (walletSigner: ethers.Wallet) => {
          try {
            const { connector, callRequest } = get()
            if (!callRequest) {
              throw new Error('No call request')
            }

            if (!connector) {
              throw new Error('No connector')
            }

            const verifyAddressMatch = (address: string) => {
              if (address.toLowerCase() !== walletSigner.address.toLowerCase()) {
                throw new Error('Address mismatch')
              }
            }

            let transaction: ethers.providers.TransactionRequest
            let dataToSign: string
            let result: string

            switch (callRequest.method) {
              case 'personal_sign': {
                dataToSign = callRequest.params[0] as string
                verifyAddressMatch(callRequest.params[1] as string)

                result = await walletSigner.signMessage(
                  ethers.utils.isHexString(dataToSign) ? ethers.utils.arrayify(dataToSign) : dataToSign
                )
                break
              }

              case 'eth_sign': {
                dataToSign = callRequest.params[1] as string
                verifyAddressMatch(callRequest.params[0] as string)

                result = await walletSigner.signMessage(
                  ethers.utils.isHexString(dataToSign) ? ethers.utils.arrayify(dataToSign) : dataToSign
                )
                break
              }

              case 'eth_signTransaction': {
                transaction = callRequest.params[0] as ethers.providers.TransactionRequest
                verifyAddressMatch(transaction.from as string)

                if ('gas' in transaction) {
                  // @ts-ignore
                  transaction.gasLimit = transaction.gas
                  // @ts-ignore
                  delete transaction.gas
                }

                result = await walletSigner.signTransaction(transaction)
                break
              }

              case 'eth_sendTransaction': {
                transaction = callRequest.params[0] as ethers.providers.TransactionRequest
                verifyAddressMatch(transaction.from as string)

                if ('gas' in transaction) {
                  // @ts-ignore
                  transaction.gasLimit = transaction.gas
                  // @ts-ignore
                  delete transaction.gas
                }

                result = (await walletSigner.sendTransaction(transaction)).hash
                break
              }

              case 'wallet_switchEthereumChain': {
                const { chainId } = callRequest.params[0] as { chainId: number }
                result = await walletSigner.signMessage(chainId.toString())
                break
              }

              default:
                throw new Error('Method not supported')
            }

            connector.approveRequest({ id: callRequest.id, result })
          } catch (error) {
            throw new Error(error.message)
          } finally {
            set({ callRequest: undefined })
          }
        },

        rejectCallRequest: () => {
          const { connector, callRequest } = get()
          if (!callRequest) {
            throw new Error('No  call request')
          }

          if (!connector) {
            throw new Error('No connector')
          }

          connector.rejectRequest({ id: callRequest.id, error: { message: 'User rejected request' } })

          set({ callRequest: undefined })
        },

        setSelectedChainId: (chainId: number) => {
          const { connector, session } = get()
          if (connector && session && session.connected) {
            connector.updateSession({ accounts: session.accounts, chainId: chainId })
          }

          set({ selectedChainId: chainId })
        },

        disconnect: async () => {
          const { connector } = get()
          if (!connector) {
            throw new Error('Connector is not initialized')
          }

          try {
            await connector.killSession()
          } catch (error) {
            console.error(error)
          } finally {
            set(initialState)

            const walletConnect = localStorage.getItem('walletconnect')
            if (walletConnect) {
              localStorage.removeItem('walletconnect')
            }
          }
        },

        clear: () => {
          set(initialState)
        },
      }),
      {
        name: 'wallet-connect-storage',
        getStorage: () => chromeSyncStorage,
      }
    ),
    { name: 'Shuttle - Wallet' }
  )
)

export const getWalletSigner = async (chainId: number, address: string, password: string) => {
  const account = Object.values(useSettings.getState().accounts)
    .map((a) => ({
      ...a,
      address: a.address.toLowerCase(),
    }))
    .find((a) => a.address === address.toLowerCase())

  if (!account) {
    throw new Error('Account not found')
  }

  try {
    const provider = getNetworkProviderByChainId(chainId)
    const wallet = await ethers.Wallet.fromEncryptedJson(account.encryptedWallet, password)
    return wallet.connect(provider)
  } catch (error) {
    throw new Error('Wrong password')
  }
}
