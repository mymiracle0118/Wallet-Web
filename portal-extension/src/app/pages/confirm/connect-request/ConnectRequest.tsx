import React, { useState } from 'react'
import { useTranslation } from 'react-i18next'
import { goBack } from 'lib/woozie'

import SinglePageTitleLayout from 'layouts/single-page-layout/SinglePageLayout'
import { CustomTypography, Button } from 'components'

import { useWalletConnect } from '@portal/shared/hooks/useWalletConnect'
import { useSettings } from '@portal/shared/hooks/useSettings'

import AccountItem from './AccountItem'
import { SpinnerIcon } from '@src/app/components/Icons'

const ConnectRequest = () => {
  const { t } = useTranslation()
  const { sessionRequest, approveSession, rejectSession, clear } = useWalletConnect()
  const { accounts } = useSettings()

  const [selectedAccount, setSelectedAccount] = useState(Object.keys(accounts)[0])
  const [loading, setLoading] = useState<boolean>(false)

  const handleApproveClick = () => {
    try {
      setLoading(true)
      approveSession([selectedAccount])
      goBack()
    } catch (error) {
      console.error('handleApproveClick', error)
    } finally {
      setLoading(false)
    }
  }

  const handleDenyClick = () => {
    try {
      setLoading(true)
      rejectSession()
    } catch (error) {
      console.error('handleDenyClick', error)
      clear()
    } finally {
      setLoading(false)
      goBack()
    }
  }

  if (!sessionRequest) {
    return null
  }

  return (
    <SinglePageTitleLayout disableBack>
      <div className="p-4 ">
        <div className="flex gap-4">
          {sessionRequest.params[0].peerMeta.icons[0] && (
            <img
              className="h-9 w-9 rounded-full"
              alt="token-thumbnail"
              src={sessionRequest.params[0].peerMeta.icons[0]}
            />
          )}
          {sessionRequest.params[0].peerMeta.name && (
            <CustomTypography className="text-h2">{sessionRequest.params[0].peerMeta.name}</CustomTypography>
          )}
        </div>
        <CustomTypography className="text-link mt-4" color="text-primary dark:text-secondary">
          {sessionRequest.params[0].peerMeta.url}
        </CustomTypography>
        <CustomTypography className="text-h1 my-4">{t('Confirm.connectRequest')}</CustomTypography>
        <CustomTypography className="text-body line leading-5">{t('Confirm.connectRequestContent')}</CustomTypography>
        <div className="my-4">
          {Object.values(accounts).map((account) => (
            <AccountItem
              key={account.address}
              selected={account.address === selectedAccount}
              accountName={account.username}
              onSelected={() => setSelectedAccount(account.address)}
            />
          ))}
        </div>
        <div className="flex gap-2  ">
          <Button
            isLoading={loading}
            variant="bordered"
            color="outlined"
            onClick={handleDenyClick}
            spinner={<SpinnerIcon />}
          >
            {t('Actions.deny')}
          </Button>
          <Button isLoading={loading} onClick={handleApproveClick} spinner={<SpinnerIcon />} color="primary">
            {!loading && t('Actions.allow')}
          </Button>
        </div>
      </div>
    </SinglePageTitleLayout>
  )
}

export default ConnectRequest
