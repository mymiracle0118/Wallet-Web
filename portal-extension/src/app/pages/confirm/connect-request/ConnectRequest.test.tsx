import React from 'react'
import { render, screen, waitFor } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import { QueryClient, QueryClientProvider } from '@tanstack/react-query'
import { I18nextProvider } from 'react-i18next'
import i18n from '@portal/shared/i18n/config'

import * as useSettingsHooks from '@portal/shared/hooks/useSettings'
import * as useWalletConnectHooks from '@portal/shared/hooks/useWalletConnect'

import ConnectRequest from 'pages/confirm/connect-request/ConnectRequest'

const mockAccountAddress = '0xcea68AdfE6D57Ff021EE9E999364b3Fb3B99Ad8d'
const mockSettingsHook = {
  accounts: {
    [mockAccountAddress]: {
      username: 'account-1',
      address: mockAccountAddress,
    },
  },
}

const mockWalletConnectHook = {
  approveSession: jest.fn(),
  rejectSession: jest.fn(),
  sessionRequest: {
    id: '',
    method: '',
    params: [
      {
        peerMeta: {
          name: 'WalletConnect',
          icons: ['https://walletconnect.org/walletconnect-logo.png'],
          url: 'https://walletconnect.org',
        },
      },
    ],
  },
}

const queryClient = new QueryClient()
const wrapper = (
  <QueryClientProvider client={queryClient}>
    <I18nextProvider i18n={i18n}>
      <ConnectRequest />
    </I18nextProvider>
  </QueryClientProvider>
)

describe('<ConnectRequest />', () => {
  beforeEach(() => {
    jest.spyOn(useSettingsHooks, 'useSettings').mockImplementation(() => mockSettingsHook)
    jest.spyOn(useWalletConnectHooks, 'useWalletConnect').mockImplementation(() => mockWalletConnectHook)
  })

  test('renders correctly', () => {
    const { asFragment } = render(wrapper)
    expect(asFragment()).toMatchSnapshot()
  })

  test('should show 1 account', async () => {
    render(wrapper)
    expect(await screen.findByText('account-1')).toBeInTheDocument()
  })

  test('should call approve session', async () => {
    render(wrapper)

    await waitFor(() => userEvent.click(screen.getByText('Allow')))
    expect(mockWalletConnectHook.approveSession).toBeCalledWith([mockAccountAddress])
  })

  test('should call reject session', async () => {
    render(wrapper)
    await waitFor(() => userEvent.click(screen.getByText('Deny')))
    expect(mockWalletConnectHook.rejectSession).toBeCalled()
  })
})
