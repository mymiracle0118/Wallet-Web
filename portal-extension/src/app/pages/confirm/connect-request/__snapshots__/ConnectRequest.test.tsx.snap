// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ConnectRequest /> renders correctly 1`] = `
<DocumentFragment>
  <div
    class="dark:bg-surface-dark bg-custom-white flex flex-col h-[600px] dark:text-custom-white text-custom-black"
  >
    <div
      class="flex flex-row relative justify-between items-center h-16"
    />
    <div
      class="w-full h-full overflow-auto"
    >
      <div
        class="p-4 "
      >
        <div
          class="flex gap-4"
        >
          <img
            alt="token-thumbnail"
            class="h-9 w-9 rounded-full"
            src="https://walletconnect.org/walletconnect-logo.png"
          />
          <div
            class="text-h2 text-custom-black dark:text-custom-white"
          >
            WalletConnect
          </div>
        </div>
        <div
          class="text-link mt-4 text-primary dark:text-secondary"
        >
          https://walletconnect.org
        </div>
        <div
          class="text-h1 my-4 text-custom-black dark:text-custom-white"
        >
          Connect request
        </div>
        <div
          class="text-body line leading-5 text-custom-black dark:text-custom-white"
        >
          By granting permission, you allow this site to view the addresses of your permitted accounts. Only connect with sites you trust.
        </div>
        <div
          class="my-4"
        >
          <div
            class="w-full flex items-center p-4 -mx-4 hover:dark:bg-custom-white10 hover:bg-custom-grey10 shadow-sm dark:shadow-custom-white10 shadow-custom-grey10"
          >
            <div
              class="flex items-center gap-2 rounded-full"
            >
              <input
                checked=""
                class="relative w-6 h-6 appearance-none border-2 border-solid border-[transparent] rounded-[4px] bg-custom-grey/10 cursor-pointer checked:before:bg-check"
                id="checkbox"
                type="checkbox"
                value=""
              />
            </div>
            <div
              class="text-h4 text-custom-black dark:text-custom-white"
            >
              account-1
            </div>
          </div>
        </div>
        <div
          class="flex gap-2  "
        >
          <button
            class="flex items-center justify-center rounded-md hover:cursor-pointer font-extrabold text-sm w-full h-[44px] border-[1px] border-solid text-primary dark:text-custom-white border-primary dark:border-custom-white40 dark:hover:border-custom-white hover:border-primary-alt hover:text-primary-alt"
            type="button"
          >
            Deny
          </button>
          <button
            class="flex items-center justify-center rounded-md hover:cursor-pointer font-extrabold text-sm w-full h-[44px] text-custom-white dark:text-custom-white hover:bg-primary-alt dark:hover:bg-primary-alt bg-primary"
            type="button"
          >
            Allow
          </button>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
